service/frontend/workflowHandler.go

```diff
@@ -3274,35 +3298,63 @@ func (wh *WorkflowHandler) createPollForDecisionTaskResponse(
 }
 
 func verifyHistoryIsComplete(
-	history *gen.History,
+	events []*gen.HistoryEvent,
 	expectedFirstEventID int64,
 	expectedLastEventID int64,
+	isFirstPage bool,
+	isLastPage bool,
+	pageSize int,
 ) error {
 
-	firstEventID := int64(-1)
-	lastEventID := int64(-1)
-	events := history.GetEvents()
 	nEvents := len(events)
-	if nEvents > 0 {
-		firstEventID = events[0].GetEventId()
-		lastEventID = events[nEvents-1].GetEventId()
+	if nEvents == 0 {
+		if isLastPage {
+			// we seem to be returning a non-nil pageToken on the lastPage which
+			// in turn cases the client to call getHistory again - only to find
+			// there are no more events to consume - bail out if this is the case here
+			return nil
+		}
+		return fmt.Errorf("invalid history: contains zero events")
+	}
+
+	firstEventID := events[0].GetEventId()
+	lastEventID := events[nEvents-1].GetEventId()
+
+	if !isFirstPage { // atleast one page of history has been read previously
+		if firstEventID <= expectedFirstEventID {
+			// not first page and no events have been read in the previous pages - not possible
+			return fmt.Errorf(
```

Comment:
Nit: use internalServiceError as our convention. Today there is only special case that we want to use fmt.Errorf because of our internal SLO calculation. I am going to clean i up with https://github.com/uber/cadence/issues/1128